include/deal.II/sundials/n_vector.templates.h

```diff
@@ -570,8 +570,8 @@ namespace SUNDIALS
         (void)v;
         return nullptr;
 #  else
-        // We need to cast away const here, as SUNDIALS demands a pure `void *`.
-        return &(const_cast<MPI_Comm &>(get_communicator<VectorType>(v)));
+        static MPI_Comm comm = get_communicator<VectorType>(v);
```

Comment:
```suggestion
        MPI_Comm comm = get_communicator<VectorType>(v);
```
or keep the existing implementation.

Reply:
@masterleinad Without static you would return a pointer to a local variable? 

@stefanozampini I think SUNDIALS just forces us to return a `MPI_Comm*` (as a `void*`) so I guess you cannot apply the change to this function. It is only internal, so maybe that is not such a big problem.

Reply:
Right... It might be best to leave this file alone.

Reply:
This file cannot be left alone since the `get_mpi_communicator` no longer returns the horrific `reference to MPI_Comm`. The solutions are:

- the current solution if the vector communicator is not supposed to change during the lifetime of an N_vector
- use a new private class member `resultcomm` to store the intermediate result

A note on this: this is happening because SUNDIALS do not follow MPI standards. `nullptr` should be `MPI_COMM_NULL`.

Reply:
Can't you let the N_Vector related `get_mpi_communicator` function still return the reference? It is an internal function and should not confuse users, I hope. I have not followed the discussion on the "why" of this change.

Reply:
No, because the get_communicator of the internal vector no longer returns a reference.

Reply:
I think `static` does not work if you would use different vectors that use different commmunicators and then call the getter for both?  So I guess, you need to add a class member in `NVectorView`.

Reply:
Will do, thanks

Reply:
What was decided here? The problem is that following this patch, all vectors simply return their communicator by-value. We no longer have place where that communicator object is stored, but this function requires us to return a pointer to such a place. The workaround with `static` only works for cases where all vectors in a program happen to use the same communicator -- if they don't, the `static` variable is initialized to the very first communicator we encounter, and that's it. 

A slightly better design would have
```
  static MPI_Comm comm;
  comm = get_communicator<VectorType>(v);
```
which at least overwrites the object every time the function is called, though this is of course neither thread-safe nor correct. In the end, this is just a boneheaded design in SUNDIALS.

I *think* what @sebproell is suggesting (nice solution, by the way!) is that the `NVectorContent` class gains another variable of type `MPI_Comm` and the `NVectorContent::get_communicator()` function returns a reference to this variable. The question is then just where exactly this member variable is initialized. But that seems simple: Because `NVectorContent` is just a wrapper around existing vectors, and because there is no other way to wrap a vector than through the constructor of `NVectorContent`, the right approach is to initialize the new member in the three constructors of this class.

@sebproell Is that how you were thinking about it?

@stefanozampini Do you want to make that change, or would you like me to take a stab at that in a separate PR?

Reply:
See also https://github.com/LLNL/sundials/issues/275

Reply:
@bangerth What you described was my idea as well (and was actually already suggested by @stefanozampini as well ;)

I mixed up the names and talked about `NVectorView` while `NVectorContent` would be correct. I would add the member and getter as you described and directly call that getter on the (`reinterpret_cast`ed) content of an NVector inside `NVectorOperations::get_communicator_as_void_ptr`.

Reply:
@bangerth Please go ahead and fix it the way you like it.  Either merge this and open a new PR or just force-push here. Thanks

Reply:
rebased on latest master for the sundials fixes